credentials/google/google.go

```diff
@@ -0,0 +1,120 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package google
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/alts"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal"
+)
+
+const tokenRequestTimeout = 30 * time.Second
+
+var vmOnGCP bool
+
+func init() {
+	vmOnGCP = alts.IsRunningOnGCP()
+}
+
+// NewDefaultCredentials returns a credentials bundle that is configured to work
+// with google services.
+//
+// This API is experimental.
+func NewDefaultCredentials() credentials.Bundle {
+	return new()
+}
+
+// creds implements credentials.Bundle.
+type creds struct {
+	// Supported modes are defined in internal/internal.go.
+	mode string
+	// The transport credentials associated with this bundle.
+	transportCreds credentials.TransportCredentials
+	// The per RPC credentials associated with this bundle.
+	perRPCCreds credentials.PerRPCCredentials
+}
+
+func (c *creds) TransportCredentials() credentials.TransportCredentials {
+	if c == nil {
+		return nil
+	}
+	return c.transportCreds
+}
+
+func (c *creds) PerRPCCredentials() credentials.PerRPCCredentials {
+	if c == nil {
+		return nil
+	}
+	return c.perRPCCreds
+}
+
+// SwitchMode should make a copy of Bundle, and switch mode. Modifying the
+// existing Bundle may cause races.
+func (c *creds) SwitchMode(mode string) (credentials.Bundle, error) {
+	if c == nil {
+		return nil, nil
+	}
+
+	newCreds := &creds{mode: mode}
+
+	// Create transport credentials.
+	switch mode {
+	case internal.CredsBundleModeTLS:
+		newCreds.transportCreds = credentials.NewTLS(nil)
+	case internal.CredsBundleModeALTS, internal.CredsBundleModeGRPCLB:
+		if !vmOnGCP {
+			return nil, errors.New("google default creds: ALTS, as part of google default credentials, is only supported on GCP")
+		}
+		// Only the clients can use google default credentials, so we only need
+		// to create new ALTS client creds here.
+		newCreds.transportCreds = alts.NewClientCreds(alts.DefaultClientOptions())
+	default:
+		return nil, fmt.Errorf("google default creds: unsupported mode: %v", mode)
+	}
+
+	// Create per RPC credentials.
+	// For the time being, we required per RPC credentials for both TLS and
+	// ALTS. In the future, this will only be required for TLS.
+	ctx, cancel := context.WithTimeout(context.Background(), tokenRequestTimeout)
+	defer cancel()
+	var err error
+	newCreds.perRPCCreds, err = oauth.NewApplicationDefault(ctx)
```

Comment:
I am pretty sure c++ strips call credentials (Oauth creds) when communicating with the balancer.

cc @ejona86 @zhangkun83 Could you pls confirm Java case?

Reply:
Skipped for GRPCLB.

Reply:
Java does not currently have a way to set CallCredentials in ManagedChannelBuilder. So there's naturally no way it could be used when communicating with the balancer.